domain/port/state/state.go

```diff
@@ -0,0 +1,271 @@
+FROM port_range
+JOIN protocol ON port_range.protocol_id = protocol.id
+WHERE unit_endpoint_uuid = $unitEndpointUUID.uuid
+`, portRange{}, unitEndpointUUID)
+
+	if err != nil {
+		return nil, errors.Annotate(err, "preparing select opened ports statement")
+	}
+
+	err = tx.Query(ctx, selectOpenedPorts, unitEndpointUUID).GetAll(&result)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return []portRange{}, nil
+	}
+	return result, domain.CoerceError(err)
+}
+
+// UpdateUnitPorts opens and closes ports for a given unit. Ports are grouped by endpoint.
+//
+// NOTE: This method opens ports first, and then closes ports. However, it is recommended
+// that your service layer reconciles overlapping port ranges before calling this method.
+func (st *State) UpdateUnitPorts(ctx context.Context, unitUUID string, openPorts, closePorts network.GroupedPortRanges) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	endpointsUnderAction := set.NewStrings()
+	for endpoint := range openPorts {
+		endpointsUnderAction.Add(endpoint)
+	}
+	for endpoint := range closePorts {
+		endpointsUnderAction.Add(endpoint)
+	}
+
+	getProtocolID, err := st.Prepare("SELECT &protocolID.* FROM protocol WHERE protocol = $protocolName.protocol", protocolID{}, protocolName{})
+	if err != nil {
+		return errors.Annotate(err, "preparing get protocol ID statement")
+	}
+
+	insertPortRange, err := st.Prepare("INSERT INTO port_range (*) VALUES ($unitPortRange.*)", unitPortRange{})
+	if err != nil {
+		return errors.Annotate(err, "preparing insert port range statement")
+	}
+
+	clearPortRange, err := st.Prepare(`
+DELETE FROM port_range
+WHERE unit_endpoint_uuid = $unitEndpointUUID.uuid
+`, unitEndpointUUID{})
+	if err != nil {
+		return errors.Annotate(err, "preparing clear port range statement")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		unitEndpointMap := map[string]unitEndpointUUID{}
+		for _, endpoint := range endpointsUnderAction.Values() {
+			var err error
+			unitEndpointMap[endpoint], err = st.getUnitEndpointUUID(ctx, tx, unitUUID, endpoint)
+			if errors.Is(err, UnitEndpointNotFound) {
+				unitEndpointMap[endpoint], err = st.insertUnitEndpoint(ctx, tx, unitUUID, endpoint)
+			}
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+
+		for endpoint, unitEndpointUUID := range unitEndpointMap {
+			currentOpenedPorts, err := st.getOpenedPorts(ctx, tx, unitEndpointUUID)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			currentNetworkPorts := transform.Slice(currentOpenedPorts, decodePortRange)
+			network.SortPortRanges(currentNetworkPorts)
+
+			resultantOpenPorts := reconcilePorts(currentNetworkPorts, openPorts[endpoint], closePorts[endpoint])
+			network.SortPortRanges(resultantOpenPorts)
+
+			// Return early if no changes are to be made. We previously sorted so
+			// port range permutations aren't picked up as changes
+			if reflect.DeepEqual(currentNetworkPorts, resultantOpenPorts) {
```

Comment:
This is a pretty heavy function, can you not use something from your interval package here?

Reply:
Removed